src/lib/mgba/websocket-client.ts

```diff
@@ -0,0 +1,558 @@
+/**
+ * WebSocket client for connecting to mGBA's eval API
+ * Handles communication with the mGBA Lua HTTP server's WebSocket endpoint
+ */
+
+import { WebSocket } from 'ws'
+import { EventEmitter } from 'events'
+
+export interface MgbaEvalResponse {
+  result?: unknown
+  error?: string
+}
+
+export interface MemoryRegion {
+  address: number
+  size: number
+  data: Uint8Array
+  lastUpdated: number
+  dirty: boolean
+}
+
+export interface SharedBufferConfig {
+  maxCacheSize: number
+  cacheTimeout: number // in milliseconds
+  preloadRegions: Array<{ address: number, size: number }>
+}
+
+export class MgbaWebSocketClient extends EventEmitter {
+  private ws: WebSocket | null = null
+  private connected = false
+
+  // Shared buffer system for memory caching
+  private readonly memoryCache = new Map<string, MemoryRegion>()
+  private sharedBufferConfig: SharedBufferConfig = {
+    maxCacheSize: 50, // Max number of cached regions
+    cacheTimeout: 5000, // 5 seconds
+    preloadRegions: [
+      { address: 0x20244e9, size: 7 }, // Party count + some context
+      { address: 0x20244ec, size: 600 }, // Full party data (6 * 100 bytes)
+    ],
+  }
```

Comment:
this system needs to support close to realtime data fetching. its important to be synced with the gamestate